### PR TITLE
feat(flags): Add relative date operators

### PR DESCRIPTION
**Pull request by neilkakkar**
## Problem

This PR adds targeting by distinct id, group keys, and relative date targeting.

~~I feel I've introduced more hacks here than the last year combined, particularly because all of these are edge-casy things: all apply only to flags, so others shouldn't show them yet, and try to reuse standard components as much as possible at the same time.~~

Things seem much nicer now, although there are still some adjustments I've made to existing components to make this work nicely.

Along the way, fix some small bugs with the RollingDateRange component.

(@Twixes , @mariusandra   - adding you for review if you know a better way to get these things done, since touches taxonomicFilters and the like)

See demo: https://posthog.slack.com/archives/C034XD440RK/p1705426858833459 (too large for github :P )

TODO in follow up:

1. [x] Backend validate relative date op values
2. [ ] Add frontend tests
3. [ ] Figure out autocomplete for current distinct id, and group key, if possible
4. [ ] Make RollingDateRange onBlur/onSelect nicer, this is annoying for both insights & now flags - need to click on 'in the last' for it to take effect.

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: frontend/src/lib/taxonomy.tsx [hunk 1]
```diff
@@ -592,6 +592,11 @@ export const KEY_MAPPING: KeyMappingInterface = {
             examples: ['16ff262c4301e5-0aa346c03894bc-39667c0e-1aeaa0-16ff262c431767'],
             system: true,
         },
+        $current_distinct_id: {
```
[the remark this answers is not in the document]
Comment (mariusandra) [reply frontend/src/lib/taxonomy.tsx:595]: please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please please let's not add any "magic properties"

File: frontend/src/lib/utils.tsx [hunk 2]
```diff
@@ -238,6 +238,11 @@ export const dateTimeOperatorMap: Record<string, string> = {
     is_not_set: '✕ is not set',
 }
 
+export const relativeDateTimeOperatorMap: Record<string, string> = {
+    is_relative_date_before: '< relative date before',
+    is_relative_date_after: '> relative date after',
```
Comment (mariusandra) [inline frontend/src/lib/utils.tsx:243]: Why not let the user select "time before" or even just "time range", and then either select a specific date or a range? The existing insight date picker could even work:

image

File: frontend/src/lib/taxonomy.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (neilkakkar) [reply frontend/src/lib/taxonomy.tsx:595]: > Why not do "has distinct ID" instead?

The goal is indeed to target distinct IDs that are just currently in use, for example: targeting an anonymous ID currently on screen, and updating flags just for that.


@mariusandra -> how do I do this without magic properties?

File: frontend/src/lib/utils.tsx (hunk 2 above)
Comment (neilkakkar) [reply frontend/src/lib/utils.tsx:243]: heh, I did not think about this, but makes a lot of sense? Will update, we _should_ be able to detect based on whether it's a timestamp or a relative date time.

File: frontend/src/lib/taxonomy.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (neilkakkar) [reply frontend/src/lib/taxonomy.tsx]: Discussed in slack, there's a new pill now

File: frontend/src/lib/components/DateFilter/DateFilter.tsx [hunk 3]
```diff
@@ -115,7 +147,17 @@ export function DateFilter({
...
 
                     return (
-                        <Tooltip key={key} title={makeLabel ? makeLabel(dateValue) : undefined}>
+                        <Tooltip
+                            key={key}
+                            title={
+                                makeLabel
+                                    ? makeLabel(
+                                          dateValue,
+                                          dateFilterToText(values[0], undefined, '', [], false, 'MMMM D, YYYY', true)
```
Comment (mariusandra) [inline frontend/src/lib/components/DateFilter/DateFilter.tsx:156]: Nit: `const whatThisIs = ...;` above and then `makeLabel(dateValue, whatThisIs)` here

**Review by mariusandra — COMMENTED** [review]
Hola, nice going! Though it now feels like this should be split up in three.

PR 1:
- The first PR should just add a metadata column with only `distinct_id` in it.
- Adding about 3 lines of code to `hogql/property.py`'s `property_to_expr` will make this metadata automatically usable in a lot of places (events list, refactored insights, etc). Would be nice 🙃.

PR 2:
- I'm a bit unconvinced of the UX with regards to selecting group keys 😅. This shouldn't block us from merging the fix for distinct ids.
- I'm also not sure of the `$group_key` field. Wouldn't it be better to stick to the actual SQL fields like `$group_0`? 
image

- Even better, we already have aliases for group tables (last part of screenshot), why not add extra virtual fields like `account_key` as an alias to `$group_0`?
- This should require just ~a few lines~ [one line](https://github.com/PostHog/posthog/blob/103280fc06083710496b509ac58388bcad2c7add/posthog/hogql/database/database.py#L184) in `database.py`
- This way we could show "Organization key" in the first part of the filter, and just the keys in the second part, making the experience meh, but ok.
- An autocomplete with actual names would be great, but I'm happy to leave that as PR 4 😅 
- We can now also add "account_key", "organization_key", etc as entries into the metadata table.

PR 3: 
- The date range improvements. 
- These seem okay to me, though I'd defer a full review to someone with more recent frontend experience

**Comment by neilkakkar** [comment]
Agree with this being too much for one PR 😅 , will split into two: distinctID + groups; and relative date stuff.

> I'm also not sure of the $group_key field. Wouldn't it be better to stick to the actual SQL fields like $group_0?

This might be true in events land, but isn't true in feature flags land when trying to query groups, because there is only a group key per group, no concept of $group_0.

A problem that will soon crop up with the metadata field is that given how taxonomic filters currently work, each pill converts to properties of the same type; so not too big a fan of putting it all in metadata here.

The current way for groups gives us autocomplete and searching by names as well, so 🙌 . Although, the meh bit is trying to then add a second key to the same filter, there autocomplete doesn't really work well.

---

I'm not really looking at it from the hogql side of things, and agree the current UX for groups wouldn't work in event filtering, specially since you're dealing with multiple groups on the same event, but this imo is too different from flags to try and incorporate into one.

**Comment by mariusandra** [comment]
> Agree with this being too much for one PR 😅 , will split into two: distinctID + groups; and relative date stuff.
> 
> > I'm also not sure of the $group_key field. Wouldn't it be better to stick to the actual SQL fields like $group_0?
> 
> This might be true in events land, but isn't true in feature flags land when trying to query groups, because there is only a group key per group, no concept of $group_0.

Aha... I understand better now.

We all do share the filtering interface, so let's figure out a way forward that works for both. I'd like to one day have a metadata field that has at least `timestamp`, `event`, `distinct_id`, `person_id` as "sort-of properties" you can filter by. People have been asking for them, and the current workaround is "use hogql", which isn't ideal considering we're talking about just adding another menu item.

I might also see adding things like `$group_0` (aka first, second, etc group's key) into the list, but that hasn't been important so far.

I get that your metadata is based on a different structure, with some overlap. The question is, how can we make an interface that lets us do both? If `$group_key` is how the property is saved on the object you're comparing, I have of course nothing against you using that property in the filter. 

However then, should we a) find a way to show/hide the metadata pill's fields based on some custom logic, b) just split this into `flag_metadata` and `event_metadata`? The second option probably makes sense then, or what do you think?

> A problem that will soon crop up with the metadata field is that given how taxonomic filters currently work, each pill converts to properties of the same type; so not too big a fan of putting it all in metadata here.

There is also the `group_type_index` field on the data coming from the filter. Could this be used here? I'm not sure I fully understand.

**Comment by neilkakkar** [comment]
I'm kinda keen to keep one metadata pill here.

>  a) find a way to show/hide the metadata pill's fields based on some custom logic

Yes, I've incorporated this already in here, metadata currently will only show options loaded via `optionsFromProps`, so this wouldn't show up if you don't want it to show up. This means you could add `$group_0` to this metadata and it will work fine everywhere, and for flags as well, because flags discard these / choose not to show these.

We can later also extend this into a propertyAllowList if need be.

For `$group_key` specifically, I opted not to use metadata but the "GroupNamesPrefix" taxonomic group because it seems to work better here - and wasn't used in taxonomic filters before, so I can customise it appropriately for flags.

**Comment by neilkakkar** [comment]
Wait what, toolbar?? 👀

**Review by mariusandra — APPROVED** [review]

**Comment by neilkakkar** [comment]
ah it's legit, it's because toolbar uses the `DateFilter` component as well, which now has another fixed date option, so it's bundled too.

**Review by Twixes — COMMENTED** [review]
I think we're much closer in the UX and it's fine for a flagged version, but this will be confusing to users:

Screenshot 2024-01-24 at 15 22 51

1. There are no "overrides" happening here, that's terminology from the dashboard date range selector. In other places, such as this, we should just say "Select a value" as the empty state.
2. The tooltip is lying, because this option does not mean "January 17". It means "in the 7 days preceding the flag being evaluated" (as far as I understand), so we should say something to that effect

**Comment by neilkakkar** [comment]
> There are no "overrides" happening here, that's terminology from the dashboard date range selector. In other places, such as this, we should just say "Select a value" as the empty state.

Yep makes sense 👍 

> The tooltip is lying, because this option does not mean "January 17". It means "in the 7 days preceding the flag being evaluated" (as far as I understand), so we should say something to that effect

Hmm, very good point, but I also want some computation here as a sanity check. Maybe something like 'Matches all values before 18 Jan if evaluated today' ? ehh.

---

In either case, will fix these before rolling out, want to get this in first to test SDKs / full e2e flow.

File: frontend/src/scenes/feature-flags/FeatureFlagReleaseConditions.tsx [hunk 4]
```diff
@@ -147,48 +148,61 @@ export function FeatureFlagReleaseConditions({
...
+                                                              '',
+                                                              [],
+                                                              false,
+                                                              String(val).slice(-1) === 'h'
+                                                                  ? 'MMMM D, YYYY HH:mm:ss'
+                                                                  : 'MMMM D, YYYY',
+                                                              true
+                                                          )} )`
+                                                        : ''}
                                                 </span>
```
[the remark this answers is not in the document]
Comment (neilkakkar) [reply frontend/src/scenes/feature-flags/FeatureFlagReleaseConditions.tsx:201]: Will do

**Pull request merged**